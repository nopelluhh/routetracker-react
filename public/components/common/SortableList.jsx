import React, { Component } from 'react'
import PropTypes from 'prop-types'
import R from 'ramda'
import { Row, Col } from 'reactstrap'

class SortableList extends Component {
	state = {
		items: []
	}

	static propTypes = {
		className: PropTypes.string,
		component: PropTypes.func.isRequired,
		list: PropTypes.array.isRequired,
		options: PropTypes.array,
		comparator: PropTypes.func
	}

	componentDidMount() {
		this.ListItem = this.props.component
		this.comparator = this.props.comparator || ((a, b) => a === b)

		this.setState({
			current: this.props.list
		})

		if (this.props.options) {
			const options = R.differenceWith(this.comparator, this.props.options, this.props.list)
			this.setState({
				options
			})
		}
	}

	componentWillReceiveProps(nextProps) {
		this.setState({
			current: nextProps.list
		})
		if (nextProps.options && this.state.current || nextProps.list) {
			const options = R.differenceWith(this.comparator, nextProps.options, nextProps.list)
			this.setState({
				options
			})
		}
	}

	render() {
		if (!this.state.current) return null
		return (
            <Row>
              <Col>
              <ul className={ this.props.className } style={ { fontSize: '1.5em' } }>
                { this.state.current.map((item, ind, arr) => 
					<this.ListItem data-source="list" key={ item.key || ind } {...item} {...this.makeHandlers(ind, arr, 'list')}/>) }
              </ul>
              </Col>
              { this.state.options ? (
                <Col>
                <ul className={ this.props.className } style={ { fontSize: '1.5em' } }>
                  { this.state.options.map((item, ind, arr) => 
						<this.ListItem data-source="options" key={ item.key || ind } {...item} {...this.makeHandlers(ind, arr, 'options')}/>
                    )}
                </ul>
                </Col>) : null }
            </Row>
		)
	}

	makeHandlers = (ind, arr, source) => ({
		draggable: true,
		onDragStart: e => {
			e.dataTransfer.setData('ind', ind)
			e.dataTransfer.setData('source', source)
			e.dataTransfer.effectAllowed = 'move'
		},
		onDragOver: e => {
			e.preventDefault()
		},
		onDrop: e => {
			let source = e.dataTransfer.getData('source')
			let dest = e.currentTarget.dataset.source

			if (source === dest) return

			let oldInd = +e.dataTransfer.getData('ind')
			let newInd = ind

			let lists = {
				list: [...this.state.current],
				options: [...this.state.options]
			}

			if (newInd == arr.length - 1) {
				newInd = arr.length
			}
			let newEl = lists[source].splice(oldInd, 1)[0]
			lists[dest].splice(newInd, 0, newEl)

			this.setState({
				current: lists.list,
				options: lists.options
			})
		}
	})
}

export default SortableList