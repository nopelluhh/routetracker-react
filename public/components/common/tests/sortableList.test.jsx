import SortableList from '../sortableList'

import React from 'react'
import renderer from 'react-test-renderer'




describe('SortableList tests', () => {
    const list = [{
        name: 'Noah'
    },
    {
        name: 'Greg'
    }]

    const Item = (props) => (
                <div>
                  {props.name }
                </div>
            )
        


    const options = [{
        name: 'Noah'
    }, {
        name: 'Greg'
    }, {
        name: 'Paul'
    }, {
        name: 'Smith'
    }, {
        name: 'Jeeves'
    },]


    it('should render a list', () => {
        const component = renderer.create(
            <SortableList list={ list } component={ Item } />
        )

        let tree = component.toJSON()
        expect(tree).toMatchSnapshot()
    })

    it('should render options when provided', () => {
        const component = renderer.create(
            <SortableList list={ list } options={options} component={ Item } />
        )

        let tree = component.toJSON()
        expect(tree).toMatchSnapshot()
    })
})
