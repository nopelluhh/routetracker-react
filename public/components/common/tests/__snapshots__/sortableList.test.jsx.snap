// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortableList tests should render a list 1`] = `
<div
  className="row"
>
  <div
    className="col"
  >
    <ul
      className={undefined}
      style={
        Object {
          "fontSize": "1.5em",
        }
      }
    >
      <div>
        Noah
      </div>
      <div>
        Greg
      </div>
    </ul>
  </div>
</div>
`;

exports[`SortableList tests should render options when provided 1`] = `
<div
  className="row"
>
  <div
    className="col"
  >
    <ul
      className={undefined}
      style={
        Object {
          "fontSize": "1.5em",
        }
      }
    >
      <div>
        Noah
      </div>
      <div>
        Greg
      </div>
    </ul>
  </div>
  <div
    className="col"
  >
    <ul
      className={undefined}
      style={
        Object {
          "fontSize": "1.5em",
        }
      }
    >
      <div>
        Noah
      </div>
      <div>
        Greg
      </div>
      <div>
        Paul
      </div>
      <div>
        Smith
      </div>
      <div>
        Jeeves
      </div>
    </ul>
  </div>
</div>
`;
